java/server/src/org/openqa/selenium/grid/sessionmap/jdbc/JdbcBackedSessionMap.java

```diff
@@ -0,0 +1,184 @@
+import org.openqa.selenium.grid.config.ConfigException;
+import org.openqa.selenium.grid.data.Session;
+import org.openqa.selenium.grid.log.LoggingOptions;
+import org.openqa.selenium.grid.sessionmap.SessionMap;
+import org.openqa.selenium.internal.Require;
+import org.openqa.selenium.json.Json;
+import org.openqa.selenium.remote.SessionId;
+import org.openqa.selenium.remote.tracing.Tracer;
+
+import java.io.Closeable;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.logging.Logger;
+
+
+public class JdbcBackedSessionMap extends SessionMap implements Closeable {
+
+  private static final Json JSON = new Json();
+  private static final Logger LOG = Logger.getLogger(JdbcBackedSessionMap.class.getName());
+  private static final String separator = "@";
+  private final String  tableName;
+  private final Connection connection;
+  private final String sessionIdCol;
+  private final String sessionCapsCol;
+
+
+  public JdbcBackedSessionMap(Tracer tracer, Connection jdbcConnection, String sessionTableName, String sessionIdCol, String sessionCapsCol)  {
+    super(tracer);
+
+    Require.nonNull("JDBC Connection Object", jdbcConnection);
+    Require.nonNull("Table name", sessionTableName);
+    Require.nonNull("Session id column name", sessionIdCol);
+    Require.nonNull("Session caps column name", sessionCapsCol);
+
+    this.connection = jdbcConnection;
+    this.tableName = sessionTableName;
+    this.sessionIdCol = sessionIdCol;
+    this.sessionCapsCol = sessionCapsCol;
+  }
+
+  public static SessionMap create(Config config) {
+    Tracer tracer = new LoggingOptions(config).getTracer();
+    JdbcSessionMapOptions sessionMapOptions = new JdbcSessionMapOptions(config);
+    String tableName = sessionMapOptions.getJdbcTableName();
+    String sessionIdColName = sessionMapOptions.getJdbcSessionIdColName();
+    String sessionCapsColName = sessionMapOptions.getJdbcSessionCapColName();
+    Connection connection;
+
+    try {
+      connection = sessionMapOptions.getJdbcConnection();
+    } catch (SQLException e) {
+      throw new ConfigException(e.toString());
+    }
+
+    return new JdbcBackedSessionMap(tracer, connection, tableName, sessionIdColName, sessionCapsColName);
+  }
+  @Override
+  public boolean add(Session session) {
+    Require.nonNull("Session to add", session);
+
+    try {
+      return insertSessionStatement(session).executeUpdate() >= 1;
+
+    } catch (SQLException e) {
+      throw new JdbcException(e.getMessage());
+    }
+  }
+
+  @Override
+  public Session get(SessionId id) throws NoSuchSessionException {
+    Require.nonNull("Session ID", id);
+
+    URI uri = null;
+    Capabilities caps = null;
+
+    try {
+      ResultSet sessions = readSessionStatement(id).executeQuery();
+
+      while(sessions.next()) {
```

Comment:
How many results are you expecting? Perhaps use a `LIMIT 1` to ensure only one row at most is sent back?

Reply:
With the limit in place, you either have 0 or 1 columns. You _should_ be able to replace the `while` with an `if`, which may simplify this a little, as you can fast-path out by throwing the `NoSuchSessionException` if there's no session found.